### PR TITLE
Weblab

### PR DESCRIPTION
**Pull request by jeremydstone**
Web Lab initial merge

### CODE REVIEW

File: apps/src/weblab/WebLab.js [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+  });
+};
+
+/**
+ * Reset WebLab to its initial state.
+ * @param {boolean} ignore Required by the API but ignored by this
+ *     implementation.
+ */
+WebLab.prototype.reset = function (ignore) {
+  // TODO - implement
```
Comment (joshlory) [inline apps/src/weblab/WebLab.js:210]: TODO?

File: dashboard/app/views/weblab_host/index.html.haml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+%html{spellcheck: 'false'}
+  %head
+  %body{style: 'margin: 0'}
+    %div{style: 'height: 20px'}
+    #bramble{style: 'height: 100vh;overflow: hidden;'}
+    %script{src: '../blockly/js/bramble/thirdparty/require.min.js', 'data-main':'../blockly/media/weblab/bramblehost.js'}
```
Comment (joshlory) [inline dashboard/app/views/weblab_host/index.html.haml]: We should probably use [`asset_path` or `minifiable_asset_path`](https://github.com/code-dot-org/code-dot-org/blob/ef9aaa7386de016ed9b484daa057f5a22a24e87c/dashboard/app/views/levels/_apps_dependencies.html.haml#L24) here.

File: apps/src/code-studio/initApp/project.js [hunk 3]
```diff
@@ -352,7 +352,9 @@ var projects = module.exports = {
         // Autosave every AUTOSAVE_INTERVAL milliseconds
         $(window).on(events.appInitialized, function () {
           // Get the initial app code as a baseline
-          currentSources.source = this.sourceHandler.getLevelSource(currentSources.source);
+          this.sourceHandler.getLevelSource(currentSources.source).then(function (response) {
```
Comment (Bjvanminnen) [inline apps/src/code-studio/initApp/project.js]: this could be an arrow function, i.e.

```
this.sourceHandler.getLevelSource(currentSources.source).then(response => {
```

File: apps/src/code-studio/initApp/project.js [hunk 4]
```diff
@@ -460,9 +464,12 @@ var projects = module.exports = {
       callback = args[0];
       forceNewVersion = args[1];
       this.sourceHandler.getAnimationList(animations => {
-        const source = this.sourceHandler.getLevelSource();
-        const html = this.sourceHandler.getLevelHtml();
-        this.save({source, html, animations}, callback, forceNewVersion);
+        var thisProject = this;
+        this.sourceHandler.getLevelSource().then(function (response) {
```
Comment (Bjvanminnen) [inline apps/src/code-studio/initApp/project.js]: this could be an arrow function, which also removes the need to capture `this` as `thisProject`:

```
this.sourceHandler.getLevelSource().then(response => {
  const source = response;
  const html = thisProject.sourceHandler.getLevelHtml();
  this.save({source, html, animations}, callback, forceNewVersion);
})
```

**Comment by joshlory** [comment]
Probably best to remove the changes to `dsls.en.yml` and the nonessential `.level` files (otherwise they're likely to trigger conflicts).

File: apps/src/code-studio/initApp/project.js [hunk 5]
```diff
@@ -535,16 +542,19 @@ var projects = module.exports = {
...
     this.sourceHandler.getAnimationList(animations => {
-      const source = this.sourceHandler.getLevelSource();
-      const html = this.sourceHandler.getLevelHtml();
-      const newSources = {source, html, animations};
-      if (JSON.stringify(currentSources) === JSON.stringify(newSources)) {
-        hasProjectChanged = false;
-        return;
-      }
+      var thisProject = this;
+      this.sourceHandler.getLevelSource().then(function (response) {
```
Comment (Bjvanminnen) [inline apps/src/code-studio/initApp/project.js]: same comment about arrow function/thisProject

File: apps/src/weblab/WebLab.js [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+import { Provider } from 'react-redux';
+
+/**
+ * An instantiable WebLab class
+ */
+
+// Global singleton
+let webLab_ = null;
+
+let WebLab = function () {
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLab.js]: could be const

File: apps/src/weblab/WebLab.js [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+  this.skin = null;
+  this.level = null;
+
+  /** @type {StudioApp} */
+  this.studioApp_ = null;
+
+  this.eventHandlers = {};
+  this.Globals = {};
+  this.reportPreloadEventHandlerComplete_ = null;
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLab.js]: the above three items might not be used anywhere?

File: apps/src/weblab/WebLab.js (hunk 6 above)
Comment (joshlory) [inline apps/src/weblab/WebLab.js]: `const`?

File: apps/src/weblab/WebLab.js [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+  config.noInstructionsWhenCollapsed = true;
+
+  let onMount = function () {
+    this.setupReduxSubscribers(this.studioApp_.reduxStore);
+
+    // TODO: understand if we need to call studioApp
+    // Other apps call studioApp.init(). That sets up UI that is not present Web Lab (run, show code, etc) and blows up
+    // if we call it. It's not clear there's anything in there we need, although we may discover there is and need to refactor it
+    // this.studioApp_.init(config);
+  }.bind(this);
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLab.js]: Could be const. Could also use an arrow function, and not need to bind this

```
const onMount = () => {
  this.setupReduxSubscribers(this.studioApp_.reduxStore);
};
```

File: apps/src/weblab/WebLab.js [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+  // Push initial level properties into the Redux store
+  this.studioApp_.setPageConstants(config, {
+    channelId: config.channel,
+    isProjectLevel: !!config.level.isProjectLevel,
+  });
+
+  function onUndo() {
+      this.brambleHost.undo();
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLab.js]: extra indent

File: apps/src/weblab/WebLab.js [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    this.brambleHost.redo();
+  }
+
+  function onShowPreview() {
+    this.brambleHost.hideTutorial();
+  }
+
+  function onShowTutorial() {
+    this.brambleHost.showTutorial();
+  }
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLab.js:131]: You might be able to create each of the above as arrow functions, and then not need to re-bind them to `this` in your JSX below (though I could be wrong about that)
i.e.

```
const onShowTutorial = () => this.brambleHost.showTutorial
```

You could actually also probably refrain from creating a function completely and in your JSX just have:

```
<WeblabView
  onShowTutorial={this.brambleHost.showTutorial}
/>
```

File: apps/src/weblab/WebLabView.jsx [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    return undefined;
+  },
+
+  componentDidMount: function () {
+    this.props.onMount();
+  },
+
+  shouldShowHeader: function () {
+    return true;
+    // return !(this.props.isEmbedView || this.props.isShareView);
```
Comment (joshlory) [inline apps/src/weblab/WebLabView.jsx]: Not used?

File: apps/src/weblab/main.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
+var appMain = require('../appMain');
+var studioApp = require('../StudioApp').singleton;
+var WebLab = require('./WebLab');
+// var blocks = require('./blocks');
+var skins = require('../skins');
```
Comment (joshlory) [inline apps/src/weblab/main.js]: Not used?

File: apps/src/weblab/WebLab.js [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+        onMount={onMount}
+      />
+    </Provider>
+  ), document.getElementById(config.containerId));
+
+  this.studioApp_.notifyInitialRenderComplete(config);
+};
+
+WebLab.prototype.getCodeAsync = function () {
+  return new Promise (function (resolve,reject) {
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLab.js]: could use an arrow func (and then no longer need to bind). also shouldn't have a space after Promise

```
return new Promise((resolve, reject) => {
```

File: apps/src/weblab/WebLab.js [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+    if (this.brambleHost !== null) {
+      this.brambleHost.getBrambleCode(function (code) {
+        resolve(code);
+      });
+    } else {
+      // Bramble not installed yet - we have no code to return
+      resolve("");
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLab.js:167]: Are we expected to hit this? Is resolving to empty string the right thing to do, vs. rejecting?

File: apps/src/weblab/WebLabView.jsx [hunk 15]
```diff
@@ -0,0 +1,73 @@
+/** @file Top-level view for GameLab */
+/* global dashboard */
+'use strict';
+
+import classNames from 'classnames';
+import {connect} from 'react-redux';
+import React from 'react';
+var _ = require('lodash');
+var StudioAppWrapper = require('../templates/StudioAppWrapper');
+var ProtectedStatefulDiv = require('../templates/ProtectedStatefulDiv');
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLabView.jsx]: can replaces these vars as imports

File: apps/src/weblab/WebLabView.jsx [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+import {isResponsiveFromState} from '../templates/ProtectedVisualizationDiv';
+import CodeWorkspace from '../templates/CodeWorkspace';
+
+/**
+ * Top-level React wrapper for WebLab
+ */
+var WebLabView = React.createClass({
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLabView.jsx]: could be const

File: apps/src/weblab/WebLabView.jsx (hunk 11 above)
Comment (Bjvanminnen) [inline apps/src/weblab/WebLabView.jsx]: remove comment?

**Comment by joshlory** [comment]
`dashboard/db/schema_cache.dump` committed — intentional?

File: apps/src/weblab/WebLabView.jsx [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+            scrolling="no"
+          />
+        </div>
+      </StudioAppWrapper>
+    );
+  }
+});
+module.exports = connect(function propsFromStore(state) {
+  return {
+  };
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLabView.jsx]: No need to connect if we're not using any of this state

File: apps/src/weblab/main.js [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+// var levels = require('./levels');
```
Comment (Bjvanminnen) [inline apps/src/weblab/main.js]: Delete comments?

Could make these require's into imports, but it might be worth leaving as is if all of our other `main.js` files look similar to this.

**Comment by joshlory** [comment]
This PR is a good candidate for rebasing into logically-grouped commits!  Optional, but it may help with tracking down where & why each file was added / modified.

File: apps/static/weblab/bramblehost.js [hunk 19]
```diff
@@ -0,0 +1,195 @@
+/**
+ * JS to communicate between Bramble and Code Studio
+ */
+
+(function () {
```
Comment (Bjvanminnen) [inline apps/static/weblab/bramblehost.js]: I'm going to see if I can find a good way to get this file into our pipeline. at a minimum the indentation in this file is wrong. If we leave it here, it's probably worth making sure it at least gets linted (I suspect it currently does not in this location)

Comment (Bjvanminnen) [reply apps/static/weblab/bramblehost.js]: If nothing else, I think we could put this in /lib and potentially get linting without sending it through webpack.

**Comment by Bjvanminnen** [comment]
When I run locally, I see a 404 for css/weblab.css  and a few other console warnings and errors.

I don't know if we expect http://localhost:3000/projects/weblab to work, but it doesn't.

File: apps/src/weblab/WebLab.js (hunk 1 above)
Comment (jeremydstone) [reply apps/src/weblab/WebLab.js:210]: Yes, to do in the future, deliberately not part of this PR.

File: apps/src/weblab/WebLab.js (hunk 14 above)
Comment (jeremydstone) [reply apps/src/weblab/WebLab.js:167]: Yes - we do hit this at level load time and resolving with empty string is correct in the context of how this is used. This is called very early to get a code baseline for autosave. The code baseline later gets updated when level sources are set. (It might be possible to arrange for this not to get called until Bramble was done loading -- but this works fine for the moment, does the right thing, and doesn't touch code outside of Web Lab.)

File: apps/src/weblab/WebLabView.jsx (hunk 17 above)
Comment (jeremydstone) [reply apps/src/weblab/WebLabView.jsx]: We will likely be using state as we add features - I'd like to leave it there as a placeholder

File: apps/src/weblab/main.js (hunk 18 above)
Comment (jeremydstone) [reply apps/src/weblab/main.js]: Good points both. Have done both. It does make Web Lab's main.js different than other apps', but we have already deliberately diverged in this PR and have ES6-ized the rest of Web Lab (in code that is very similar to other apps). So might as well go all-in on ES6 in Web Lab!

File: apps/src/weblab/WebLabView.jsx (hunk 17 above)
Comment (Bjvanminnen) [reply apps/src/weblab/WebLabView.jsx]: I would push back a little bit and say that we shouldn't connect this component to the store until we know we need to - it's easy enough to add this back in the future.

Also, just noticing now that this could/should be `export default WeblabView` instead of `module.exports = ...`

File: dashboard/app/models/pd/enrollment.rb [hunk 20]
```diff
@@ -17,7 +17,6 @@
 #  user_id             :integer
 #  survey_sent_at      :datetime
 #  completed_survey_id :integer
-#  school_info_id      :integer
```
Comment (ashercodeorg) [inline dashboard/app/models/pd/enrollment.rb:20]: This should presumably not be part of this PR.

**Comment by ashercodeorg** [comment]
Is WebLab going to require being age thirteen or teacher permission to access? If so, those privacy / security changes should be part of this PR.

File: apps/src/weblab/WebLabView.jsx (hunk 17 above)
Comment (jeremydstone) [reply apps/src/weblab/WebLabView.jsx]: All right, I am persuaded! Done (both).

File: dashboard/app/views/weblab_host/index.html.haml (hunk 2 above)
Comment (jeremydstone) [reply dashboard/app/views/weblab_host/index.html.haml]: Thx, done

**Comment by jeremydstone** [comment]
Not seeing a 404 for weblab.css (it serves it with a 200; from apps/build/package/css/weblab.css). There are some warnings and errors when running Web Lab level, all from within Bramble. Most or all of these show up in Bramble's demo page. We are OK for now to enable web lab level authoring, we will get the Bramble spew sorted out as we go along.

File: apps/src/code-studio/initApp/initApp.js [hunk 21]
```diff
@@ -240,16 +240,24 @@ window.apps = {
...
-      var source;
-      if (window.Blockly) {
-        // If we're readOnly, source hasn't changed at all
-        source = Blockly.mainBlockSpace.isReadOnly() ? currentLevelSource :
-          Blockly.Xml.domToText(Blockly.Xml.blockSpaceToDom(Blockly.mainBlockSpace));
-      } else if (appOptions.getCode) {
-        source = appOptions.getCode();
-      }
-      return source;
+      return new Promise (function (resolve,reject) {
```
Comment (Bjvanminnen) [inline apps/src/code-studio/initApp/initApp.js]: Nit, remove space after `Promise` and add space between resolve/reject. Also, this could be an arrow function if you wanted.

File: apps/src/code-studio/initApp/initApp.js [hunk 22]
```diff
@@ -240,16 +240,24 @@ window.apps = {
...
+        var source;
```
Comment (Bjvanminnen) [inline apps/src/code-studio/initApp/initApp.js]: nit: could be let

File: apps/src/code-studio/initApp/initApp.js [hunk 23]
```diff
@@ -240,16 +240,24 @@ window.apps = {
...
+        if (window.Blockly) {
+          // If we're readOnly, source hasn't changed at all
+          source = Blockly.mainBlockSpace.isReadOnly() ? currentLevelSource :
+              Blockly.Xml.domToText(Blockly.Xml.blockSpaceToDom(Blockly.mainBlockSpace));
+          resolve(source);
+        } else if (appOptions.getCode) {
+          source = appOptions.getCode();
+          resolve(source);
+        } else if (appOptions.getCodeAsync) {
+          appOptions.getCodeAsync().then(function (source) {
```
Comment (Bjvanminnen) [inline apps/src/code-studio/initApp/initApp.js]: nit could be arrow function, remove space between function and paren if not.

File: apps/src/weblab/WebLab.js [hunk 24]
```diff
@@ -0,0 +1,213 @@
...
+      this.brambleHost.enableInspector();
+    } else {
+      this.brambleHost.disableInspector();
+    }
+  }
+
+  ReactDOM.render((
+    <Provider store={this.studioApp_.reduxStore}>
+      <WebLabView
+        onUndo={onUndo.bind(this)}
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLab.js:146]: could this just be `this.brambleHost.undo.bind(this.brambleHost)` instead of creating a new local function that just calls brambleHost.undo? Same with redo.

File: apps/src/weblab/WebLab.js [hunk 25]
```diff
@@ -0,0 +1,213 @@
...
+        onMount={onMount}
+      />
+    </Provider>
+  ), document.getElementById(config.containerId));
+
+  this.studioApp_.notifyInitialRenderComplete(config);
+};
+
+WebLab.prototype.getCodeAsync = function () {
+  return new Promise((resolve,reject) => {
```
Comment (Bjvanminnen) [inline apps/src/weblab/WebLab.js]: nit: space after comma

File: apps/src/code-studio/initApp/initApp.js (hunk 23 above)
Comment (Bjvanminnen) [reply apps/src/code-studio/initApp/initApp.js]: actually, i guess we're inconsistent about space after `function` or not :-/

File: apps/src/weblab/brambleHost.js [hunk 26]
```diff
@@ -0,0 +1,195 @@
+/* global requirejs */
+
+/**
+ * JS to communicate between Bramble and Code Studio
+ */
+
+(function () {
```
Comment (Bjvanminnen) [inline apps/src/weblab/brambleHost.js]: no need for this all to be in an anon function now that it's going through webpack

File: apps/src/weblab/brambleHost.js [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+
+    // the main Bramble object -- used to access file system
+    var bramble_ = null;
+    // the Bramble proxy object -- used to access methods on the Bramble UI frame
+    var brambleProxy_ = null;
+    // interface to Web Lab host in Code Studio
+    var webLab_ = null;
+
+    // Project root in file system
+    var projectRoot = "/codedotorg/weblab";
```
Comment (Bjvanminnen) [inline apps/src/weblab/brambleHost.js]: can use let/const throughout this file now instead

File: apps/src/weblab/brambleHost.js [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+
+    function putFilesInBramble(Bramble, sources, callback) {
+        var fs = Bramble.getFileSystem();
+        var sh = new fs.Shell();
+        var Path = Bramble.Filer.Path;
```
Comment (Bjvanminnen) [inline apps/src/weblab/brambleHost.js]: indentation throughout this file should be two spaces instead of four

File: apps/src/weblab/brambleHost.js [hunk 29]
```diff
@@ -0,0 +1,195 @@
...
+
+        sh.rm(projectRoot, {recursive: true}, function (err) {
```
Comment (Bjvanminnen) [inline apps/src/weblab/brambleHost.js]: could make a lot of these callbacks arrow funcs now if you wanted, now that we're going through our JS pipeline

File: apps/src/weblab/brambleHost.js [hunk 30]
```diff
@@ -0,0 +1,195 @@
...
+                }
+
+                // start an async-chained enumeration through the file list
+                writeSourceFile(sources, 0, callback);
+            });
+        });
+    }
+
+    function getFilesFromBramble(callback) {
+
```
Comment (Bjvanminnen) [inline apps/src/weblab/brambleHost.js]: nit: empty line of whitespace

File: apps/src/weblab/WebLab.js (hunk 24 above)
Comment (jeremydstone) [reply apps/src/weblab/WebLab.js:146]: It turns out, no. At the time the React component is created, the bramble component hasn't been created yet so this.brambleHost is null, so we need the indirection to the local function. It may (or may not) be possible to further simplify this down the road.

Comment (Bjvanminnen) [reply apps/src/weblab/WebLab.js:146]: Gotcha, makes sense. Another thing you probably could do is just have an arrow function in place, i.e.

```
<WeblabView
  onUndo={() => this.brambleHost.undo()}
/>
```

Up to you which approach you like better (I don't feel super strongly either way).

**Comment by Bjvanminnen** [comment]
lgtm

**Comment by breville** [comment]
naming suggestion: give each commit a description rather than just WIP.
